File: abc417/d/main.py
# 謎の1TLE

from bisect import bisect_right
from functools import lru_cache
import sys
sys.setrecursionlimit(10**7)


N = int(input())
gifts = [tuple(int(x) for x in input().split()) for _ in range(N)]

@lru_cache(maxsize=None)
def solve(finished: int, now_point: int):
    if finished >= N:
        return now_point

    if now_point <= gifts[finished][0]:
        now_point += gifts[finished][1]
    else:
        now_point -= gifts[finished][2]
        now_point = max(now_point, 0)

    return solve(finished + 1, now_point)

down_points_sum = [0]
for x in gifts:
    down_points_sum.append(down_points_sum[-1] + x[2])

Q = int(input())
for _ in range(Q):
    X = int(input())
    if X <= 1000:
        print(solve(0, X))
        continue

    downs = bisect_right(down_points_sum, X - 1000)
    if downs >= len(down_points_sum):
        print(X - down_points_sum[-1])
        continue

    now_point = X - down_points_sum[downs]
    print(solve(downs, now_point))

File: abc417/d/main.cpp
// メモ化再帰の実装方法はしっかりチェック。
// C++の練習もした方が良いな...。

#include "../../cpp_macros/atcoder.h"

unordered_map<ll, ll> memo;
int N;
vvl gifts;


ll solve(int finished, int now_point) {
    ll key = finished * 100000LL + now_point;
    if (memo.count(key)) return memo[key];
    if (finished >= N) {
        memo[key] = now_point;
        return now_point;
    };
    if (now_point <= gifts[finished][0]) {
        now_point += gifts[finished][1];
    } else {
        now_point -= gifts[finished][2];
        now_point = max(now_point, 0);
    }

    ll result = solve(finished + 1, now_point);
    memo[key] = result;
    return result;
}

int main() {
    cin >> N;
    int p, a, b;
    rep(_, N) {
        cin >> p >> a >> b;
        gifts.push_back({p, a, b});
    }

    vl down_points_sum = {0};
    rep(i, N) {
        down_points_sum.push_back(down_points_sum[down_points_sum.size() - 1] + gifts[i][2]);
    }

    // rep(i, down_points_sum.size()) {
    //     cout << down_points_sum[i] << endl;
    // }

    int Q;
    cin >> Q;
    rep(_, Q) {
        int X;
        cin >> X;
        if (X <= 1000) {
            cout << solve(0, X) << endl;
            continue;
        }

        auto it = upper_bound(down_points_sum.begin(), down_points_sum.end(), X - 1000);
        int downs = it - down_points_sum.begin();

        if (downs >= down_points_sum.size()) {
            cout << X - down_points_sum[down_points_sum.size() - 1] << endl;
            continue;
        }

        int now_point = X - down_points_sum[downs];
        cout << solve(downs, now_point) << endl;
    }
}
